### PR TITLE
fix: kleros liquid extra views bugs

### PR DESCRIPTION
**Pull request by fnanni-0**
These fixes were already implemented in the xDAI contract:
- https://github.com/kleros/xdai-kleros-liquid/pull/15
- https://github.com/kleros/xdai-kleros-liquid/pull/17

### CODE REVIEW

File: test/kleros/kleros-liquid-extra-views.js [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+      subcourtTree.timesPerPeriod,
+      subcourtTree.sortitionSumTreeK
+    )
+
+    const extraData = `0x${(0).toString(16).padStart(64, '0')}${(1)
+      .toString(16)
+      .padStart(64, '0')}`
+    await klerosLiquid.createDispute(2, extraData, {
+      value: await klerosLiquid.arbitrationCost(extraData)
+    })
```
Comment (hbarcelos) [inline test/kleros/kleros-liquid-extra-views.js:138]: Do we really need to create dispute to pass the phase?

File: test/kleros/kleros-liquid-extra-views.js [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      subcourtTree.children[1].minStake
+    )
+    await klerosLiquid.setStake(
+      subcourtTree.children[0].children[0].ID,
+      subcourtTree.children[0].children[0].minStake
+    )
+    klerosLiquid.setStake(
+      subcourtTree.children[0].children[1].ID,
+      subcourtTree.children[0].children[1].minStake
+    )
```
Comment (hbarcelos) [inline test/kleros/kleros-liquid-extra-views.js:219]: Is this call meant to fail? That's why you don't `await` it?

File: test/kleros/kleros-liquid-extra-views.js [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    })
+    await klerosLiquid.setStake(1, subcourtTree.minStake + 1, {
+      from: governor
+    })
+
+    const stake0 = await extraViews.stakeOf(governor, 0)
+    const stake1 = await extraViews.stakeOf(governor, 1)
+    expect(stake0).to.eql(web3.toBigNumber(subcourtTree.minStake))
+    expect(stake1).to.eql(web3.toBigNumber(subcourtTree.minStake + 1))
+
```
Comment (hbarcelos) [inline test/kleros/kleros-liquid-extra-views.js:153]: This test case could be split in two to improve readability.
I'd suggest to create a nested suite with `describe` like this:

```javascript
describe('When there are delayed stakes', () => {
  beforeEach('Setup delayed stakes', () => {
    // ...
  });

  it('Should get the total amount of tokens staked in a subcourt', () => {
    // Tests for .stakeOf()
  });

  it('Should get the proper juror stats', () => {
    // Tests for .getJuror()
  });
});
```

File: test/kleros/kleros-liquid-extra-views.js [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    expect(jurorData[2]).to.eql(web3.toBigNumber(0))
+    expect(jurorData[3]).to.deep.equal([
+      stake0,
+      stake1,
+      web3.toBigNumber(0),
+      web3.toBigNumber(0)
+    ])
+  })
+
+  it('Should ignore invalid delayed stakes.', async () => {
```
Comment (hbarcelos) [inline test/kleros/kleros-liquid-extra-views.js:171]: I don't get the purpose of this test at all.
Transactions leading to invalid stakes would just be reverted, wouldn't they?

**Review by hbarcelos — CHANGES_REQUESTED** [review]
The test cases are a bit hard to follow because it's written in very low level.

For methods returning tuples, prefer to use the named parameters instead of the positional ones, so they have meaning.

Also I don't see a regression test that reproduces the original problem that you identified.
It would be great if you could write a failing test for the previous version and make it pass in this version.

File: test/kleros/kleros-liquid-extra-views.js (hunk 1 above)
Comment (fnanni-0) [reply test/kleros/kleros-liquid-extra-views.js]: Yes, because of this https://github.com/kleros/kleros/blob/dae4e63f6e386421dd8ead22411d3be8ba4500ad/contracts/kleros/KlerosLiquid.sol#L393.

File: test/kleros/kleros-liquid-extra-views.js (hunk 3 above)
Comment (fnanni-0) [reply test/kleros/kleros-liquid-extra-views.js:148]: Yes, I changed it.

File: test/kleros/kleros-liquid-extra-views.js (hunk 4 above)
Comment (fnanni-0) [reply test/kleros/kleros-liquid-extra-views.js]: No, in KlerosLiquid, the only invalid delayed stake that would revert is trying to stake in a non-existing subcourt. All other invalid cases fail silently when delayed stakes are executed https://github.com/kleros/kleros/blob/master/contracts/kleros/KlerosLiquid.sol#L464.

Comment (fnanni-0) [reply test/kleros/kleros-liquid-extra-views.js]: The view contract should take this into account.

**Comment by fnanni-0** [comment]
> Also I don't see a regression test that reproduces the original problem that you identified.
> It would be great if you could write a failing test for the previous version and make it pass in this version.

Can you point me to which of the 4 problems are you referring? I think all of them are considered in the tests (although I see that not all of them are explicit).

**Comment by hbarcelos** [comment]
> > Also I don't see a regression test that reproduces the original problem that you identified.
> > It would be great if you could write a failing test for the previous version and make it pass in this version.
> 
> Can you point me to which of the 4 problems are you referring? I think all of them are considered in the tests (although I see that not all of them are explicit).

Oh, I see. Maybe try making it explicitly then, what do you think?

**Comment by fnanni-0** [comment]
> Oh, I see. Maybe try making it explicitly then, what do you think?

Yeah, I see what you mean, but I'm finding it hard to do it, because the issues where very basic. The first (very simple) test that checks if the stats returned are those expected, already covers the 3 bugs mentioned here https://github.com/kleros/xdai-kleros-liquid/pull/15.

File: test/kleros/kleros-liquid-extra-views.js (hunk 4 above)
Comment (hbarcelos) [reply test/kleros/kleros-liquid-extra-views.js]: Got it.

**Review by hbarcelos — APPROVED** [review]
LGTM now.
